packages/runtime/container-runtime/src/test/gcUnreferencedStateTracker.spec.ts

```diff
@@ -0,0 +1,146 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/* eslint-disable max-len */
+
+import { strict as assert } from "assert";
+import { SinonFakeTimers, useFakeTimers } from "sinon";
+import {
+    UnreferencedState,
+    UnreferencedStateTracker,
+} from "../garbageCollection";
+
+describe("Garbage Collection Tests", () => {
+    let clock: SinonFakeTimers;
+
+    before(() => {
+        clock = useFakeTimers();
+    });
+
+    afterEach(() => {
+        clock.reset();
+    });
+
+    after(() => {
+        clock.restore();
+    });
+
+    describe("UnreferencedStateTracker", () => {
+        /**
+         * During the lifetime of an unreferenced object, its state is tracked and updated in two ways.
+         * Timers are set to trigger transitioning to the next state, and updateTracking is also called
+         * whenever GC runs.
+         * These tests specify how to advance the clock (to hit the timers) and also when to call updateTracking,
+         * checking that the expected state transitions occur as specified
+         */
+        function runTestCase(testCase: { start: [number, UnreferencedState]; steps: [number, number | undefined, UnreferencedState][]; }) {
+            return () => {
+                const [startTimestamp, startState] = testCase.start;
+                const tracker = new UnreferencedStateTracker(
+                    0 /* unreferencedTimestampMs */,
+                    10 /* inactiveTimeoutMs */,
+                    20 /* sweepTimeoutMs */,
+                    startTimestamp /* currentReferenceTimestampMs */,
+                );
+                assert.equal(tracker.state, startState, `Wrong starting state`);
+                let lastTime = startTimestamp;
+                testCase.steps.forEach(([localTime, currentReferenceTimestampMs, expectedState], index) => {
+                    assert(localTime > lastTime, "INVALID TEST CASE: steps must move forward in time, following start");
+                    const delta = localTime - lastTime;
+                    clock.tick(delta);
+                    lastTime = localTime;
+
+                    if (currentReferenceTimestampMs !== undefined) {
+                        tracker.updateTracking(currentReferenceTimestampMs);
+                    }
+
+                    assert.equal(tracker.state, expectedState, `Wrong state at step ${index}`);
+                });
+                tracker.stopTracking();
+            };
+        }
+
+        // In all cases:  unreferencedTimestampMs = 0, inactiveTimeoutMs = 10, sweepTimeoutMs = 20
+        const testCases: { name: string; start: [number, UnreferencedState]; steps: [number, number | undefined, UnreferencedState][]; }[] = [
```

Comment:
Note: It took a little time to understand the whole test cases thing. Especially the steps, maybe a comment or two. I'm assuming this is what it is for each step `[localTime, currentReferenceTimestampMs, expectedState]`. It feels like each step is a test case, and each test case is a test. I'm not sure what the best way to name it is.

Reply:
Yeah I'll add a bigger comment.  The steps are points in time for a given tracked node, to simulate calls to updateTracking and to let the timers fire.  Each test case walks through different scenarios of when those each happen.  There's also a comment on `runTestCase` that explains some of it.